apiserver/test/integration/clientset_test.go

```diff
@@ -1349,3 +1355,109 @@ func testIPAMConfigClient(client calicoclient.Interface, name string) error {
 
 	return nil
 }
+
+// TestBlockAffinityClient exercises the BlockAffinity client.
+func TestBlockAffinityClient(t *testing.T) {
+	const name = "test-blockaffinity"
+	rootTestFunc := func() func(t *testing.T) {
+		return func(t *testing.T) {
+			client, shutdownServer := getFreshApiserverAndClient(t, func() runtime.Object {
+				return &v3.BlockAffinity{}
+			})
+			defer shutdownServer()
+			if err := testBlockAffinityClient(client, name); err != nil {
+				t.Fatal(err)
+			}
+		}
+	}
+
+	if !t.Run(name, rootTestFunc()) {
+		t.Errorf("test-blockaffinity test failed")
+	}
+}
+
+func testBlockAffinityClient(client calicoclient.Interface, name string) error {
+	blockAffinityClient := client.ProjectcalicoV3().BlockAffinities()
+	blockAffinity := &v3.BlockAffinity{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+
+		Spec: v3.BlockAffinitySpec{
+			CIDR:  "10.0.0.0/24",
+			Node:  "node1",
+			State: "pending",
+		},
+	}
+	libV3BlockAffinity := &libapiv3.BlockAffinity{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+
+		Spec: libapiv3.BlockAffinitySpec{
+			CIDR:    "10.0.0.0/24",
+			Node:    "node1",
+			State:   "pending",
+			Deleted: "false",
+		},
+	}
+	ctx := context.Background()
+
+	// Calico libv3 client instantiation in order to get around the API create restrictions
+	// TODO: Currently these tests only run on a Kubernetes datastore since profile creation
+	// does not work in etcd. Figure out how to divide this configuration to etcd once that
+	// is fixed.
+	config := apiconfig.NewCalicoAPIConfig()
+	config.Spec = apiconfig.CalicoAPIConfigSpec{
+		DatastoreType: apiconfig.Kubernetes,
+		EtcdConfig: apiconfig.EtcdConfig{
+			EtcdEndpoints: "http://localhost:2379",
+		},
+		KubeConfig: apiconfig.KubeConfig{
+			Kubeconfig: os.Getenv("KUBECONFIG"),
+		},
+	}
+	apiClient, err := libclient.New(*config)
+	if err != nil {
+		return fmt.Errorf("unable to create Calico lib v3 client: %s", err)
+	}
+
+	_, err = blockAffinityClient.Create(ctx, blockAffinity, metav1.CreateOptions{})
+	if err == nil {
+		return fmt.Errorf("should not be able to create block affinity %s ", blockAffinity.Name)
+	}
+
+	// Create the block affinity using the libv3 client.
+	_, err = apiClient.BlockAffinities().Create(ctx, libV3BlockAffinity, options.SetOptions{})
+	if err != nil {
+		return fmt.Errorf("error creating the object through the Calico v3 API '%v' (%v)", libV3BlockAffinity, err)
+	}
+
+	blockAffinityNew, err := blockAffinityClient.Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("error getting object %s (%s)", name, err)
+	}
+
+	blockAffinityList, err := blockAffinityClient.List(ctx, metav1.ListOptions{})
```

Comment:
The primary use-case for this API involves watches, so I think we want a test that ensures Watches behave correctly as well.